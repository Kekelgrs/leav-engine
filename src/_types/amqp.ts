export enum RoutingKeys {
    FILES_EVENT = 'files.event',
    FILES_PREVIEW_REQUEST = 'files.previewRequest',
    FILES_PREVIEW_RESPONSE = 'files.previewResponse'
}

export interface IMessageBody {
    [key: string]: any;
}

export type onMessageFunc = (msg: string) => Promise<void>;
