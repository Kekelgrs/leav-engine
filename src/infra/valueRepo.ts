import {IAttribute, AttributeTypes} from '../domain/attributeDomain';
import {IValue} from '../domain/valueDomain';
import {IDbService} from './db/dbService';
import {aql} from 'arangojs';

const VALUES_COLLECTION = 'core_values';
const VALUES_LINKS_COLLECTION = 'core_edge_values_links';

export interface IValueRepo {
    saveValue?(library: string, recordId: string, attribute: IAttribute, value: IValue): Promise<IValue>;
    getValueById?(valueId: number): Promise<IValue>;
    getValues?(library: string, recordId: number, attribute: IAttribute): Promise<IValue[]>;
}

export default function(dbService: IDbService | any): IValueRepo {
    return {
        async saveValue(library: string, recordId: string, attribute: IAttribute, value: IValue): Promise<IValue> {
            if (attribute.type === AttributeTypes.INDEX) {
                const collec = dbService.db.collection(library);
                const updatedDoc = await collec.update(
                    {
                        _key: recordId
                    },
                    {
                        [attribute.id]: value.value
                    }
                );

                const docData = await collec.document(updatedDoc);

                return {value: docData[attribute.id]};
            } else if (attribute.type === AttributeTypes.STANDARD) {
                const valCollec = dbService.db.collection(VALUES_COLLECTION);
                const edgeCollec = dbService.db.edgeCollection(VALUES_LINKS_COLLECTION);

                const valueData = {
                    value: value.value
                };

                const savedVal = value.id
                    ? await valCollec.update({_key: value.id}, valueData)
                    : await valCollec.save(valueData);

                const savedValDoc = await valCollec.document(savedVal);

                const edgeData = {
                    _from: library + '/' + recordId,
                    _to: savedVal._id,
                    attribute: attribute.id,
                    modified_at: value.modified_at,
                    created_at: value.created_at
                };

                let savedEdge;
                if (value.id) {
                    await edgeCollec.updateByExample({_from: edgeData._from, _to: edgeData._to}, edgeData);
                    savedEdge = await edgeCollec.firstExample({_from: edgeData._from, _to: edgeData._to});
                } else {
                    savedEdge = await edgeCollec.save(edgeData);
                    savedEdge = await edgeCollec.firstExample(savedEdge);
                }

                return {
                    id: savedValDoc._key,
                    value: savedValDoc.value,
                    attribute: savedEdge.attribute,
                    modified_at: savedEdge.modified_at,
                    created_at: savedEdge.created_at
                };
            }
        },
        async getValueById(valueId: number): Promise<IValue> {
            const valCollec = dbService.db.collection(VALUES_COLLECTION);
            const edgeCollec = dbService.db.edgeCollection(VALUES_LINKS_COLLECTION);

            const values = await valCollec.lookupByKeys([valueId]);

            if (!values.length) {
                return null;
            }

            const valueLinks = await edgeCollec.inEdges(values[0]);

            return {
                id: values[0]._key,
                value: values[0].value,
                attribute: valueLinks[0].attribute,
                modified_at: valueLinks[0].modified_at,
                created_at: valueLinks[0].created_at
            };
        },
        async getValues(library: string, recordId: number, attribute: IAttribute): Promise<IValue[]> {
            if (attribute.type === AttributeTypes.INDEX) {
                const res = await dbService.execute(aql`
                    FOR r IN ${library}
                        FILTER r._key == ${recordId}
                        RETURN r.${attribute.id}
                `);

                return [
                    {
                        value: res[0],
                        attribute: attribute.id
                    }
                ];
            } else if (attribute.type === AttributeTypes.STANDARD) {
                const res = await dbService.execute(aql`
                    FOR value, edge
                        IN 1 OUTBOUND ${library + '/' + recordId}
                        ${VALUES_LINKS_COLLECTION}
                        FILTER edge.attribute == ${attribute.id}
                        RETURN {value, edge}
                `);

                return res.map(r => ({
                    id: r.value._key,
                    value: r.value.value,
                    attribute: r.edge.attribute,
                    modified_at: r.edge.modified_at,
                    created_at: r.edge.created_at
                }));
            }
        }
    };
}
