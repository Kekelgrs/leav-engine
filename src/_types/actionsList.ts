import {description} from 'joi';
import {IAttribute} from './attribute';
import libraryDomain from 'domain/libraryDomain';

export enum ActionsListEvents {
    SAVE_VALUE = 'saveValue',
    DELETE_VALUE = 'deleteValue',
    GET_VALUE = 'getValue'
}

export enum ActionsListIOTypes {
    STRING = 'string',
    NUMBER = 'number',
    OBJECT = 'object',
    BOOLEAN = 'boolean'
}

export type ActionsListValueType = string | number | boolean | {};

// TODO: should be able to use [event: ActionsListEvents] as from TS v2.9
export interface IActionsListConfig {
    [ActionsListEvents.SAVE_VALUE]?: [IActionsListSavedAction];
    [ActionsListEvents.DELETE_VALUE]?: [IActionsListSavedAction];
    [ActionsListEvents.GET_VALUE]?: [IActionsListSavedAction];
}

export interface IActionsListContext {
    attribute?: IAttribute;
    library?: string;
    recordId?: number;
}

export interface IActionsListParams {
    [name: string]: any;
}

export interface IActionsListFunction {
    name: string;
    description: string;
    inputTypes: ActionsListIOTypes[];
    outputTypes: ActionsListIOTypes[];
    params?: [{name: string; type: string; description: string}];
    action: (value: ActionsListValueType, params: IActionsListParams, ctx: IActionsListContext) => ActionsListValueType;
}

export interface IActionsListSavedAction {
    name: string;
    isSystem: boolean;
    params?: [{name: string; value: string}];
}
