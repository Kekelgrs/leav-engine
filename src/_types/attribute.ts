import {IActionsListConfig} from './actionsList';
import {ITreePermissionsConf} from './permissions';
import {ISystemTranslation} from './systemTranslation';

export interface IAttribute {
    id: string;
    system?: boolean;
    label?: ISystemTranslation;
    type: AttributeTypes;
    format?: AttributeFormats;
    linked_library?: string;
    linked_tree?: string;
    embedded_fields?: IEmbeddedAttribute[];
    actions_list?: IActionsListConfig;
    permissions_conf?: ITreePermissionsConf;
    multiple_values?: boolean;
    versions_conf?: IAttributeVersionsConf;
}

export enum ValueVersionMode {
    SIMPLE = 'simple',
    SMART = 'smart'
}

export interface IAttributeVersionsConf {
    versionable: boolean;
    mode?: ValueVersionMode;
    trees?: string[];
}

export interface IEmbeddedAttribute {
    id: string;
    label?: ISystemTranslation;
    format?: AttributeFormats;
    validation_regex?: string;
    embedded_fields?: IEmbeddedAttribute[];
}

/**
 * Accepted fields to filter attributes list
 */
export interface IAttributeFilterOptions {
    id?: string;
    type?: AttributeTypes[];
    format?: AttributeFormats[];
    system?: boolean;
    label?: string;
    linked_library?: string;
    multiple_values?: boolean;
}

export enum AttributeTypes {
    SIMPLE = 'simple',
    SIMPLE_LINK = 'simple_link',
    ADVANCED = 'advanced',
    ADVANCED_LINK = 'advanced_link',
    TREE = 'tree'
}

export enum AttributeFormats {
    TEXT = 'text',
    NUMERIC = 'numeric',
    DATE = 'date',
    ENCRYPTED = 'encrypted',
    BOOLEAN = 'boolean',
    EXTENDED = 'extended'
}

export enum IOTypes {
    STRING = 'string',
    NUMBER = 'number',
    BOOLEAN = 'boolean',
    OBJECT = 'object'
}
