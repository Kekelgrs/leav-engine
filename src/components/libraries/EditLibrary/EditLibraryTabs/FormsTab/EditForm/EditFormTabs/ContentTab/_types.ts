import {DragObjectWithType} from 'react-dnd';
import {GET_ATTRIBUTES_attributes_list} from '../../../../../../../../_gqlTypes/GET_ATTRIBUTES';
import {FormElementTypes} from '../../../../../../../../_gqlTypes/globalTypes';
import {IKeyValue} from '../../../../../../../../_types/shared';

export const PLACEHOLDER_ID = '__placeholder__';

export enum UIElementTypes {
    DIVIDER = 'divider',
    FIELDS_CONTAINER = 'fields_container',
    TEXT_BLOCK = 'text_block',
    TABS = 'tabs'
}

export enum FieldTypes {
    TEXT_INPUT = 'input_field',
    DATE = 'date',
    CHECKBOX = 'checkbox',
    ENCRYPTED = 'encrypted',
    DROPDOWN = 'dropdown'
}

export enum DraggableElementTypes {
    RESERVE_LAYOUT_ELEMENT = 'RESERVE_LAYOUT_ELEMENT',
    ATTRIBUTE = 'ATTRIBUTE',
    FORM_ELEMENT = 'FORM_ELEMENT'
}

export interface IFormElement {
    id: string;
    order: number;
    type: FormElementTypes;
    uiElement: IUIElement;
    containerId: string;
    settings?: IKeyValue<any>;
    herited?: boolean;
}

export interface IUIElement {
    type: UIElementTypes | FieldTypes;
    component: JSX.Element;
    canDrop: (dropCandidate: IFormElement) => boolean;
    settings?: string[];
}

export interface IFormElementProps<T extends object> {
    elementData?: IFormElement;
    settings: T;
}

export interface ICommonFieldsSettings {
    label?: string;
    attribute?: GET_ATTRIBUTES_attributes_list;
}

export interface IFormElementPos {
    order: number;
    containerId: string;
}

export interface IFormBuilderDragObject<T extends IUIElement | IFormElement> extends DragObjectWithType {
    type: DraggableElementTypes;
    element: T;
    index: number;
    currentPos?: IFormElementPos;
    originPos?: IFormElementPos;
    dropAtPos?: IFormElementPos;
}
