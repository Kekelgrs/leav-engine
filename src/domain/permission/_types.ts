import {
    AppPermissionsActions,
    IPermissionsTreeTarget,
    LibraryPermissionsActions,
    PermissionsActions,
    PermissionTypes
} from '_types/permissions';
import {IQueryInfos} from '_types/queryInfos';
import {ITreeNode} from '_types/tree';

export interface IPermissionTarget {
    attributeId?: string;
    recordId?: string;
}

export interface IGetSimplePermissionsParams {
    type: PermissionTypes;
    applyTo: string | null;
    action: PermissionsActions;
    usersGroupId: string;
    permissionTreeTarget?: IPermissionsTreeTarget;
    ctx: IQueryInfos;
}

export interface IGetPermissionsByActionsParams {
    type: PermissionTypes;
    applyTo: string | null;
    actions: PermissionsActions[];
    usersGroupId: string;
    permissionTreeTarget?: IPermissionsTreeTarget;
    ctx: IQueryInfos;
}

export type PermByActionsRes = {[name: string]: boolean | null} | null;

export interface IGetPermissionByUserGroupsParams {
    type: PermissionTypes;
    action: PermissionsActions;
    userGroupsPaths: ITreeNode[][];
    applyTo?: string;
    permissionTreeTarget?: IPermissionsTreeTarget;
    ctx: IQueryInfos;
}

export interface IGetAppPermissionParams {
    action: AppPermissionsActions;
    userId: string;
    ctx: IQueryInfos;
}

export interface IGetHeritedAppPermissionParams {
    action: AppPermissionsActions;
    userGroupId: string;
    ctx: IQueryInfos;
}

export interface IGetLibraryPermissionParams {
    action: LibraryPermissionsActions;
    libraryId: string;
    userId: string;
    ctx: IQueryInfos;
}

export interface IGetHeritedLibraryPermissionParams {
    action: LibraryPermissionsActions;
    libraryId: string;
    userGroupId: string;
    ctx: IQueryInfos;
}

export interface IGetHeritedPermissionsParams {
    type: PermissionTypes;
    applyTo: string;
    action: PermissionsActions;
    userGroupId: string;
    permissionTreeTarget?: IPermissionsTreeTarget;
    ctx: IQueryInfos;
}

export interface IIsAllowedParams {
    type: PermissionTypes;
    action: PermissionsActions;
    userId: string;
    applyTo?: string;
    target?: IPermissionTarget;
    ctx: IQueryInfos;
}
