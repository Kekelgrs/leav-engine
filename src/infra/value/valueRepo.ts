import {IAttribute} from '_types/attribute';
import {IValue, IValuesOptions} from '_types/value';
import {IAttributeTypesRepo} from '../attributeTypes/attributeTypesRepo';

export interface IValueRepo {
    /**
     * Create a new value
     *
     * @param library
     * @param recordId
     * @param attribute
     * @param value
     * @param {}
     */
    createValue(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue>;

    /**
     * Update an existing value. Field "id" is expected on the value
     * @param library
     * @param recordId
     * @param attribute
     * @param value
     * @param {}
     */
    updateValue(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue>;

    /**
     * Delete an existing value. Field "id" is expected on the value
     *
     * @param library
     * @param recordId
     * @param attribute
     * @param value
     * @param {}
     */
    deleteValue(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue>;

    /**
     * Get all values for given record and attribute
     *
     * @param library
     * @param recordId
     * @param attribute
     * @return Array<{}>    Return an empty array if no value found
     */
    getValues(
        library: string,
        recordId: number,
        attribute: IAttribute,
        forceGetAllValues?: boolean,
        options?: IValuesOptions
    ): Promise<IValue[]>;

    /**
     * Return a specific value based on its ID. Field "id" is expect on the value
     *
     * @param library
     * @param recordId
     * @param attribute
     * @param value
     * @return {}   Return null if no value found
     */
    getValueById(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue>;
    clearAllValues(attribute: IAttribute): Promise<boolean>;
}

interface IDeps {
    'core.infra.attributeTypes'?: IAttributeTypesRepo;
}

export default function({'core.infra.attributeTypes': attributeTypesRepo = null}: IDeps = {}): IValueRepo {
    return {
        createValue(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue> {
            const typeRepo = attributeTypesRepo.getTypeRepo(attribute);
            return typeRepo.createValue(library, recordId, attribute, value);
        },
        updateValue(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue> {
            const typeRepo = attributeTypesRepo.getTypeRepo(attribute);
            return typeRepo.updateValue(library, recordId, attribute, value);
        },
        deleteValue(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue> {
            const typeRepo = attributeTypesRepo.getTypeRepo(attribute);
            return typeRepo.deleteValue(library, recordId, attribute, value);
        },
        getValues(
            library: string,
            recordId: number,
            attribute: IAttribute,
            forceGetAllValues?: boolean,
            options?: IValuesOptions
        ): Promise<IValue[]> {
            const typeRepo = attributeTypesRepo.getTypeRepo(attribute);
            return typeRepo.getValues(library, recordId, attribute, forceGetAllValues, options);
        },
        getValueById(library: string, recordId: number, attribute: IAttribute, value: IValue): Promise<IValue> {
            const typeRepo = attributeTypesRepo.getTypeRepo(attribute);
            return typeRepo.getValueById(library, recordId, attribute, value);
        },
        clearAllValues(attribute: IAttribute): Promise<boolean> {
            const typeRepo = attributeTypesRepo.getTypeRepo(attribute);
            return typeRepo.clearAllValues(attribute);
        }
    };
}
