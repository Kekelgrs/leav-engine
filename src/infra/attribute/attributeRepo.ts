import {aql} from 'arangojs';
import {IList, IPaginationParams} from '_types/list';
import {AttributeFormats, AttributeTypes, IAttribute, IAttributeFilterOptions} from '../../_types/attribute';
import {ILibrary} from '../../_types/library';
import {IDbService} from '../db/dbService';
import {IDbUtils} from '../db/dbUtils';
import {LIB_ATTRIB_COLLECTION_NAME} from '../library/libraryRepo';
import {IValueRepo} from '../value/valueRepo';

export interface IAttributeRepo {
    getAttributes(
        filters?: IAttributeFilterOptions,
        strictFilters?: boolean,
        withCount?: boolean,
        pagination?: IPaginationParams
    ): Promise<IList<IAttribute>>;
    updateAttribute(attrData: IAttribute): Promise<IAttribute>;
    createAttribute(attrData: IAttribute): Promise<IAttribute>;
    deleteAttribute(attrData: IAttribute): Promise<IAttribute>;

    /**
     * Get all libraries for which this attribute is enabled
     *
     * @param attribute
     */
    getLibrariesUsingAttribute?(attribute: IAttribute): Promise<ILibrary[]>;
}

export const ATTRIB_COLLECTION_NAME = 'core_attributes';

export default function(
    dbService: IDbService | null = null,
    dbUtils: IDbUtils | null = null,
    valueRepo: IValueRepo | null = null,
    config = null
): IAttributeRepo {
    return {
        async getAttributes(
            filters?: IAttributeFilterOptions,
            strictFilters: boolean = false,
            withCount: boolean = false,
            pagination?: IPaginationParams
        ): Promise<IList<IAttribute>> {
            return dbUtils.findCoreEntity<IAttribute>(
                ATTRIB_COLLECTION_NAME,
                filters,
                strictFilters,
                withCount,
                pagination
            );
        },
        async updateAttribute(attrData: IAttribute): Promise<IAttribute> {
            const defaultParams = {
                _key: '',
                system: false,
                label: {fr: '', en: ''},
                type: AttributeTypes.ADVANCED,
                format: AttributeFormats.TEXT,
                multiple_values: false
            };
            let docToInsert = dbUtils.convertToDoc(attrData);

            docToInsert = {...defaultParams, ...docToInsert};

            // Insert in libraries collection
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);
            const res = await dbService.execute(aql`UPDATE ${docToInsert} IN ${col} RETURN NEW`);

            return dbUtils.cleanup(res.pop());
        },
        async createAttribute(attrData: IAttribute): Promise<IAttribute> {
            const defaultParams = {
                _key: '',
                system: false,
                label: {fr: '', en: ''},
                type: AttributeTypes.ADVANCED,
                format: AttributeFormats.TEXT,
                multiple_values: false
            };
            let docToInsert = dbUtils.convertToDoc(attrData);

            docToInsert = {...defaultParams, ...docToInsert};

            // Insert in libraries collection
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);
            const res = await dbService.execute(aql`INSERT ${docToInsert} IN ${col} RETURN NEW`);

            return dbUtils.cleanup(res.pop());
        },
        async deleteAttribute(attrData: IAttribute): Promise<IAttribute> {
            // Delete links library<->attribute
            const libAttributesCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);

            // Delete all values
            await valueRepo.clearAllValues(attrData);

            const resDelEdges = await dbService.execute(aql`
                    FOR e IN ${libAttributesCollec}
                        FILTER e._to == ${'core_attributes/' + attrData.id}
                        REMOVE e IN ${libAttributesCollec}
                    `);

            // Delete attribute
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);

            const res = await dbService.execute(aql`REMOVE ${{_key: attrData.id}} IN ${col} RETURN OLD`);

            // Return deleted attribute
            return dbUtils.cleanup(res.pop());
        },
        async getLibrariesUsingAttribute(attribute: IAttribute): Promise<ILibrary[]> {
            const libAttribCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);

            // TODO: use aql template tag, and find out why it doesn't work :)
            const query = `
                FOR v
                IN 1 INBOUND '${ATTRIB_COLLECTION_NAME}/${attribute.id}'
                ${LIB_ATTRIB_COLLECTION_NAME}
                RETURN v
            `;

            const res = await dbService.execute(query);

            return res.map(dbUtils.cleanup);
        }
    };
}
