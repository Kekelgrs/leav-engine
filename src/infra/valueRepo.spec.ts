import valueRepo from './valueRepo';
import {Database} from 'arangojs';
import {AttributeTypes} from '../_types/attribute';

describe('ValueRepo', () => {
    describe('saveValue', () => {
        test('Should save an indexed value', async function() {
            const updatedRecordData = {
                _id: 'test_lib/222435651',
                _rev: '_WSywvyC--_',
                _key: 222435651,
                test_attr: 'test_val'
            };

            const updatedValueData = {
                value: 'test_val'
            };

            const mockDbCollec = {
                update: global.__mockPromise(updatedRecordData),
                document: global.__mockPromise(updatedRecordData)
            };

            const mockDb = {collection: jest.fn().mockReturnValue(mockDbCollec)};

            const mockDbServ = {db: mockDb};

            const valRepo = valueRepo(mockDbServ);

            const createdVal = await valRepo.saveValue(
                'test_lib',
                '12345',
                {
                    id: 'test_attr',
                    type: AttributeTypes.INDEX
                },
                {
                    value: 'test val'
                }
            );

            expect(mockDbCollec.update.mock.calls.length).toBe(1);
            expect(mockDbCollec.update).toBeCalledWith({_key: '12345'}, {test_attr: 'test val'});

            expect(createdVal).toMatchObject(updatedValueData);
        });

        test('Should save a new standard value', async function() {
            const createdValueData = {
                _id: 'core_values/987654',
                _rev: '_WSywvyC--_',
                _key: 987654,
                value: 'test_val'
            };

            const createdEdgeData = {
                _id: 'core_edge_values_links/222435651',
                _rev: '_WSywvyC--_',
                _from: 'test_lib/12345',
                _to: 'core_values/987654',
                _key: 978654321,
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999
            };

            const newValueData = {
                id: 987654,
                value: 'test_val',
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999
            };

            const mockDbCollec = {
                save: global.__mockPromise(createdValueData),
                document: global.__mockPromise(createdValueData)
            };

            const mockDbEdgeCollec = {
                save: global.__mockPromise(createdEdgeData),
                firstExample: global.__mockPromise(createdEdgeData)
            };

            const mockDb = {
                collection: jest.fn().mockReturnValue(mockDbCollec),
                edgeCollection: jest.fn().mockReturnValue(mockDbEdgeCollec)
            };

            const mockDbServ = {db: mockDb};

            const valRepo = valueRepo(mockDbServ);

            const createdVal = await valRepo.saveValue(
                'test_lib',
                '12345',
                {
                    id: 'test_attr',
                    type: AttributeTypes.STANDARD
                },
                {
                    value: 'test val',
                    modified_at: 400999999,
                    created_at: 400999999
                }
            );

            expect(mockDbCollec.save.mock.calls.length).toBe(1);
            expect(mockDbCollec.save).toBeCalledWith({value: 'test val'});

            expect(mockDbEdgeCollec.save.mock.calls.length).toBe(1);
            expect(mockDbEdgeCollec.save).toBeCalledWith({
                _from: 'test_lib/12345',
                _to: 'core_values/987654',
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999
            });

            expect(createdVal).toMatchObject(newValueData);
        });

        test('Should update a standard value', async function() {
            const savedValueData = {
                _id: 'core_values/987654',
                _rev: '_WSywvyC--_',
                _key: 987654,
                value: 'test_val'
            };

            const savedEdgeData = {
                _id: 'core_edge_values_links/222435651',
                _rev: '_WSywvyC--_',
                _from: 'test_lib/12345',
                _to: 'core_values/987654',
                _key: 978654321,
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999
            };

            const valueData = {
                id: 987654,
                value: 'test_val',
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999
            };

            const mockDbCollec = {
                update: global.__mockPromise(savedValueData),
                document: global.__mockPromise(savedValueData)
            };

            const mockDbEdgeCollec = {
                updateByExample: global.__mockPromise(),
                firstExample: global.__mockPromise(savedEdgeData)
            };

            const mockDb = {
                collection: jest.fn().mockReturnValue(mockDbCollec),
                edgeCollection: jest.fn().mockReturnValue(mockDbEdgeCollec)
            };

            const mockDbServ = {db: mockDb};

            const valRepo = valueRepo(mockDbServ);

            const savedVal = await valRepo.saveValue(
                'test_lib',
                '12345',
                {
                    id: 'test_attr',
                    type: AttributeTypes.STANDARD
                },
                {
                    id: 987654,
                    value: 'test val',
                    modified_at: 500999999
                }
            );

            expect(mockDbCollec.update.mock.calls.length).toBe(1);
            expect(mockDbCollec.update).toBeCalledWith({_key: 987654}, {value: 'test val'});

            expect(mockDbEdgeCollec.updateByExample.mock.calls.length).toBe(1);
            expect(mockDbEdgeCollec.updateByExample).toBeCalledWith(
                {
                    _from: 'test_lib/12345',
                    _to: 'core_values/987654'
                },
                {
                    _from: 'test_lib/12345',
                    _to: 'core_values/987654',
                    attribute: 'test_attr',
                    modified_at: 500999999
                }
            );

            expect(savedVal).toMatchObject(valueData);
        });
    });

    describe('getValueByID', () => {
        test('Should return value', async function() {
            const lookupValueRes = [
                {
                    _key: 987654,
                    value: 'test val'
                }
            ];

            const edgeRes = [
                {
                    _from: 'test_lib/987654',
                    _to: 'core_values/987654',
                    modified_at: 99999,
                    created_at: 99999,
                    attribute: 'test_attr'
                }
            ];

            const mockDbCollec = {
                lookupByKeys: global.__mockPromise(lookupValueRes)
            };

            const mockDbEdgeCollec = {
                inEdges: global.__mockPromise(edgeRes)
            };

            const mockDb = {
                collection: jest.fn().mockReturnValue(mockDbCollec),
                edgeCollection: jest.fn().mockReturnValue(mockDbEdgeCollec)
            };

            const mockDbServ = {db: mockDb};

            const valRepo = valueRepo(mockDbServ);

            const value = await valRepo.getValueById(987654);

            expect(mockDbCollec.lookupByKeys.mock.calls.length).toBe(1);
            expect(mockDbEdgeCollec.inEdges.mock.calls.length).toBe(1);
            expect(value).toMatchObject({
                id: 987654,
                value: 'test val',
                modified_at: 99999,
                created_at: 99999,
                attribute: 'test_attr'
            });
        });

        test("Should return null if value doesn't exists", async function() {
            const mockDbCollec = {
                lookupByKeys: global.__mockPromise([])
            };

            const mockDbEdgeCollec = {
                inEdges: global.__mockPromise()
            };

            const mockDb = {
                collection: jest.fn().mockReturnValue(mockDbCollec),
                edgeCollection: jest.fn().mockReturnValue(mockDbEdgeCollec)
            };

            const mockDbServ = {db: mockDb};

            const valRepo = valueRepo(mockDbServ);

            const value = await valRepo.getValueById(987654);

            expect(mockDbCollec.lookupByKeys.mock.calls.length).toBe(1);
            expect(mockDbEdgeCollec.inEdges.mock.calls.length).toBe(0);
            expect(value).toBeNull();
        });
    });
    describe('getValues', () => {
        test('Should return values for standard attribute', async function() {
            const traversalRes = [
                {
                    value: {
                        _key: 987654,
                        value: 'test val'
                    },
                    edge: {
                        _from: 'test_lib/123456',
                        _to: 'core_values/987654',
                        attribute: 'test_attr',
                        modified_at: 99999,
                        created_at: 99999
                    }
                },
                {
                    value: {
                        _key: 987655,
                        value: 'test val2'
                    },
                    edge: {
                        _from: 'test_lib/123456',
                        _to: 'core_values/987655',
                        attribute: 'test_attr',
                        modified_at: 99999,
                        created_at: 99999
                    }
                }
            ];

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(traversalRes)
            };

            const valRepo = valueRepo(mockDbServ);

            const values = await valRepo.getValues('test_lib', 123456, {
                id: 'test_attr',
                type: AttributeTypes.STANDARD
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].bindVars).toMatchSnapshot();

            expect(values.length).toBe(2);
            expect(values[0]).toMatchObject({
                id: 987654,
                value: 'test val',
                attribute: 'test_attr',
                modified_at: 99999,
                created_at: 99999
            });
        });

        test('Should return values for index attribute', async function() {
            const queryRes = ['test val'];

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(queryRes)
            };

            const valRepo = valueRepo(mockDbServ);

            const values = await valRepo.getValues('test_lib', 123456, {
                id: 'test_attr',
                type: AttributeTypes.INDEX
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].bindVars).toMatchSnapshot();

            expect(values.length).toBe(1);
            expect(values[0]).toMatchObject({
                value: 'test val',
                attribute: 'test_attr'
            });
        });
    });
});
