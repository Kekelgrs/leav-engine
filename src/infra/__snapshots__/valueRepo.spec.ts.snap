// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueRepo getValues Should return values for index attribute 1`] = `
"
                    FOR r IN @value0
                        FILTER r._key == @value1
                        RETURN r.@value2
                "
`;

exports[`ValueRepo getValues Should return values for index attribute 2`] = `
Object {
  "value0": "test_lib",
  "value1": 123456,
  "value2": "test_attr",
}
`;

exports[`ValueRepo getValues Should return values for standard attribute 1`] = `
"
                    FOR value, edge
                        IN 1 OUTBOUND @value0
                        @value1
                        FILTER edge.attribute == @value2
                        RETURN {value, edge}
                "
`;

exports[`ValueRepo getValues Should return values for standard attribute 2`] = `
Object {
  "value0": "test_lib/123456",
  "value1": "core_edge_values_links",
  "value2": "test_attr",
}
`;
